### PR TITLE
CT-27 decoupling permissiong

### PR DESCRIPTION
**Pull request by brama051**

### CODE REVIEW

File: packages/server/src/modules/permissions/model.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    holderType: {} as unknown as DepopulatedPermission['holderType'],
+    role: {} as unknown as DepopulatedPermission['role'],
+    object: {} as unknown as DepopulatedPermission['object'],
+    objectType: {} as unknown as DepopulatedPermission['objectType'],
+    v: {} as unknown as number,
+  },
+  { minimize: false },
+);
+
+stepSchema.index({
```
Comment (MrBr) [inline packages/server/src/modules/permissions/model.ts]: What is this for?

Got a feeling this should be used in `step` module rather then permissions. Seems like a permissions tool that other modules should use?

File: packages/server/src/modules/permissions/service.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  // is assigned to
+  const permissions = await PermissionModel.find(query).lean().exec();
+  const roles = permissions.map(item => item.role);
+  return new Promise<boolean>(() =>
+    hasPermissions(roles, object.type, predicate),
+  );
+};
+
+export const getUsersWithRoleService = async (
+  object: Step | Chapter | Lesson | Activity, // | UserGroup
```
Comment (MrBr) [inline packages/server/src/modules/permissions/service.ts:60]: I'd rather say `object: Subject` to be more generic. Then, if there is a permission with such combination it will return it.

This way we'll have to extend the service each time we add a new subject type.

Comment (MrBr) [reply packages/server/src/modules/permissions/service.ts:60]: Or even change args to `id, type, role`. Idk

File: packages/server/src/modules/activity/service.ts [hunk 3]
```diff
@@ -129,12 +129,19 @@ export const findActivities = (
...
-  userId: User['id'],
-) =>
-  Array.isArray(activities)
-    ? activities
-    : [activities].every(activity =>
-        activity.roles.some(({ user }) => user.id === userId),
-      );
+  user: User,
+) => {
+  return new Promise<boolean>(function resolve() {
```
Comment (MrBr) [inline packages/server/src/modules/activity/service.ts:136]: Seems like `Promise.all` could be helpful here. Something like:

```
Promise.all([].map(async () =>  await application.service.hasPermissionToDo(
          user,
          activity,
          'participate',
        ) ))
```

File: packages/server/src/modules/activity/routes.ts [hunk 4]
```diff
@@ -45,14 +45,12 @@ application.service.registerPostRoute(
...
   conditional((req, res) => !res.locals.activity.v)(
-    toLocals('notification.users', (req, res) =>
-      res.locals.activity.roles
-        .map(
-          ({ user, role }: Role<LessonActivityRole>) =>
-            role === LessonActivityRole.STUDENT && user,
-        )
-        .filter(Boolean),
-    ),
+    toLocals('notification.users', async (req, res) => {
```
Comment (MrBr) [inline packages/server/src/modules/activity/routes.ts:48]: Does `toLocals` support async return?

File: packages/models/src/role/types.ts [hunk 5]
```diff
@@ -1,11 +1,107 @@
...
+};
+
+export const hasPermissions = (
+  roles: string[],
+  objectType: string,
+  action: string,
+): boolean => {
+  const objectTypeRoleDefinitions = getRoleDefinitionsByObjectType(objectType);
+
+  let found = false;
```
Comment (MrBr) [inline packages/models/src/role/types.ts]: Array `some` method will help you here.
```
const hasPermission = roles.some(() => objectTypeRoleDefinitions[role]?.includes(action));
```

File: packages/server/src/modules/activity/middleware.ts [hunk 6]
```diff
@@ -63,24 +63,20 @@ export const findActivities: MiddlewareFunction = (req, res, next) => {
...
-export const sendActivity: MiddlewareFunction = (req, res, next) => {
+export const sendActivity: MiddlewareFunction = async (req, res, next) => {
   const activity = res.locals.activity as Activity;
   if (!activity) {
     throw new ActivityNotPreparedError();
   }
   // TODO - verify if this is working
-  [...activity.roles]
-    .map(({ user }) => user)
+  (await application.service.getUsersWithRole(activity, 'Participator'))
```
Comment (MrBr) [inline packages/server/src/modules/activity/middleware.ts:79]: Because of the permission inheritance all users with higher level of permission will also get the message?

Comment (brama051) [reply packages/server/src/modules/activity/middleware.ts:79]: When querying for user permission on action, stronger role will inherit it as well, but when querying for users with specific role, you only get explicitly assigned. 

It's not set in stone. I'm still in the process of figuring out what's needed from the permission API.

File: packages/server/src/modules/activity/routes.ts (hunk 4 above)
Comment (brama051) [reply packages/server/src/modules/activity/routes.ts:48]: I hope that they do :D 
It crossed my mind yesterday but only briefly. It'll get obvious fast if it doesn't.

File: packages/server/src/modules/permissions/service.ts (hunk 2 above)
Comment (brama051) [reply packages/server/src/modules/permissions/service.ts:60]: There's definitely more work to do to make this strongly and openly typed.

File: packages/server/src/modules/permissions/model.ts (hunk 1 above)
Comment (brama051) [reply packages/server/src/modules/permissions/model.ts]: Obi Wan: "This wasn't copy pasted"

File: packages/server/src/modules/permissions/service.ts (hunk 2 above)
Comment (brama051) [reply packages/server/src/modules/permissions/service.ts:60]: I'm planning to remove uses of magic text when prompting for role or action.

File: packages/server/src/modules/activity/routes.ts (hunk 4 above)
Comment (MrBr) [reply packages/server/src/modules/activity/routes.ts:48]: Checked it. It supports promise.

File: packages/server/src/modules/activity/routes.ts [hunk 7]
```diff
@@ -45,14 +45,12 @@ application.service.registerPostRoute(
...
+      await application.service.getUsersWithRole(
+        res.locals.activity,
+        'Participator',
```
Comment (MrBr) [inline packages/server/src/modules/activity/routes.ts:51]: It would be better to use `enum` for this.

**Comment by MrBr** [comment]
@brama051 I'd like to a middleware implementation for the permissions. Have you thought about it?

File: packages/server/src/application/types/index.ts [hunk 8]
```diff
@@ -146,6 +144,21 @@ export type Service = {
   addUser: (user: User) => Promise<void>;
   addTag: (tag: Tag) => Promise<void>;
   findTags: (startsWith: string) => Promise<Tag[]>;
+  addPermission: (
+    object: Step | Chapter | Lesson | Activity, // | UserGroup
+    role: string,
+    holder: User, // | UserGroup
+  ) => Promise<void>;
+  hasPermissionToDo: (
```
Comment (MrBr) [inline packages/server/src/application/types/index.ts]: Use generics.

File: packages/server/src/modules/activity/middleware.ts (hunk 6 above)
Comment (MrBr) [reply packages/server/src/modules/activity/middleware.ts:79]: Ok, in this case then I think all the users associated with this subject should get activity. 

Maybe it makes sense to have a different signature which allows omitting role to get all associated users. Another option is to create a new service for that purpose.

TLDR; we want sometimes (in this case) to notify all associated users.

File: packages/server/src/modules/activity/service.ts [hunk 9]
```diff
@@ -129,12 +129,19 @@ export const findActivities = (
       });
   });
 
-export const canEditActivities = (
+export const canEditActivities = async (
```
Comment (MrBr) [inline packages/server/src/modules/activity/service.ts:132]: I think we won't need this function in the future since our permission infrastructure will be generic enough to support validation.

Comment (MrBr) [reply packages/server/src/modules/activity/service.ts:132]: On a side note, this function here presents a specific validation problem when we want to validate permissions for multiple activities.

File: packages/server/src/modules/permissions/service.ts [hunk 10]
```diff
@@ -68,3 +70,40 @@ export const getUsersWithRoleService = async (
       depopulatedPermission.toObject<Permission>().holder,
   );
 };
+export const getUserObjectsByActionService = async (
```
Comment (MrBr) [inline packages/server/src/modules/permissions/service.ts:76]: Where do you see this function used?

File: packages/models/src/permissions/types.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  objectType: String; //typeof TYPE_STEP | typeof TYPE_CHAPTER | typeof TYPE_LESSON | typeof TYPE_ACTIVITY  // | typeof TYPE_USER_GROUP
+}
+
+export interface NormalizedPermission {
+  id: Permission['id'];
+  type: Permission['type'];
+  holder: User['id']; // | UserGroup['id']
+  holderType: Permission['holderType'];
+  role: Permission['role'];
+  object: Step['id'] | Chapter['id'] | Lesson['id'] | Activity['id'];
```
Comment (MrBr) [inline packages/models/src/permissions/types.ts:26]: Can we rename object to `entity`. This is name has been used in the application for this purpose.

File: packages/server/src/application/migrations/client.ts [hunk 12]
```diff
@@ -3,8 +3,14 @@ import path from 'path';
...
-  return client.db(process.env.DB_NAME) as Db;
+  if (!client) {
+    client = new MongoClient(process.env.DB_URL as string);
+    await client.connect();
+  }
+
+  return [client.db(process.env.DB_NAME) as Db, client] as const;
 };
+
+// process.on('beforeExit', async () => await client.close());
```
Comment (MrBr) [inline packages/server/src/application/migrations/client.ts:16]: Why is this commented?

File: packages/models/src/role/types.ts [hunk 13]
```diff
@@ -1,11 +1,108 @@
...
+export const getRolesWithAction = (
+  objectType: string,
+  action: string,
+): string[] => {
+  const roleDefinitions = getRoleDefinitionsByObjectType(objectType);
+  return Object.keys(roleDefinitions).filter((role: string) =>
+    roleDefinitions[role].includes(action),
+  );
+};
+export const hasPermissions = (
```
Comment (MrBr) [inline packages/models/src/role/types.ts:204]: I'd name this `isRoleAction` or something like that. `hasPermission` looks like end product, a function that will actually resolve holder permission.

File: packages/server/src/modules/permissions/service.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    role: { $in: roles },
+  };
+
+  return (await PermissionModel.find(query).exec()).map(
+    depopulatedPermission =>
+      depopulatedPermission.toObject<DepopulatedPermission>().object,
+  );
+};
+
+export const getUserObjectsByRoleService = async (
```
Comment (MrBr) [inline packages/server/src/modules/permissions/service.ts:96]: `getUserObjectIdsByRoleService` seems more precise.

Comment (MrBr) [reply packages/server/src/modules/permissions/service.ts:96]: How will this function handle pagination? I feel like we should have an aggregation mechanism which will allow us to construct more powerful query that will resolve all this.

Please check out mongo aggregations - https://www.mongodb.com/docs/manual/core/aggregation-pipeline/

File: packages/server/src/modules/step/routes.ts [hunk 15]
```diff
@@ -6,13 +6,25 @@ import {
...
 
-const { identify, sendData, sendStatusOk, toLocals } = middleware;
+const { identify, sendData, sendStatusOk, toLocals, validatePermissions } =
+  middleware;
 
 application.service.registerPostRoute(
   '/step/save',
   identify,
   toLocals('step', req => req.body),
+  conditional(async (req, res) => {
```
Comment (MrBr) [inline packages/server/src/modules/step/routes.ts:20]: You can do this as one liner

```
  conditional(async (req, res) => await stepExists(req.body.id))(
    validatePermissions(
      req => ({ id: req.body.id, type: TYPE_STEP }),
      'updateContent',
    ),)
```
